### DISCUSSION TITLE
Enabling java community to have easier access to jbang "friendly" java projects

### DISCUSSION THREAD

**Discussion by maxandersen** [jbangdev/jbang#1733] [Ideas]
For the last few years I've been proposing various projects I saw could benefit from having a jbang-catalog.
Often the project is inside an org and I then propose to make a org wide jbang-catalog to shorten the name.

Sometime this succeeds http://github.com/microsoft/jbang-catalog (thanks @brunoborges), other times, it gets [discouraged](https://github.com/jvm-profiling-tools/ap-loader/issues/19) and sometimes even having any jbang-catalog gets [rejected](https://github.com/julianhyde/sqlline/pull/430).

Most of the discouragement or rejection falls into one of two categories:

A) having a jbang-catalog in org or repo is felt like an overhead/unclear who maintain it 
B) "no one asked for it"

Its a classical chicken-egg situation - if jbang was well-known and fully understood this would not be a problem but to become well-known and fully understood need to have it more used.

I fully get the resistance and sometime it just wont happen...but even if others don't use jbang-catalog in their project I really would like to make it easy for anyone to use other peoples great software.

I've resisted adding various commands to jbangdev central catalog as I much rather prefer each project to own their tooling experience; but I also see that for some projects its just not going to happen...so can we make a community driven approach that does not hurt or offend project owners and encourage them to adopt owning their jbang catalogs?

In a conversation with @parttimenerd he had some ideas and suggestions that made think if we could find a way to improve this.

# jbangapp/jbang-catalog 

One way of doing this would be to add aliases to a central jbang catalog. It can't be jbangdev/jbang-catalog as it is automatically blessed and I don't want to accidentally expose jbang users to a security issue; but it could be i.e. jbangapp/jbang-catalog making `jbangapp` the key name. (I'm not married to jbangapp but so far all my other ideas (jtool, javaapp, jbox, ..) was taken or had trademark issues)

Idea is that in that org we or anyone else can submit PR's to add their favorite set of tools.

i.e. `sqlline` who don't want a catalog would then be made available to anyone by using `jbang sqlline@jbangapp` or ap-profiler could shorten their `jbang jfr2flame@jvm-profiling-tools/ap-profiler` to `jbang jfr2flame@jbangapp`.

The advantages of this are that:

- its short and concise
- easy to maintain (renovate can be used to keep versions uptodate)
- its testable (could add small testsuite for each alias to check it still works between updates)
- anyone can contribute
- will enable *a lot* of tools to be instantly available to anyone with java/jbang installed, enables us to bootstrap a way to easily do what other language ecosystems has (npm install, pip install etc.)

The downsides are:

- its one big catalog thus over time it will be updated often and might slow things down for users; even for commands that have not updated
- stable versioning is problematic (if we constantly update the commands users will get a potential update without doing anything and no easy way to "lock" the version because its a global catalog rather than a smaller catalog - one workaround is that users do `jbang sqlline@jbangapp/<shaorbranchid>) 
- adding commands to this might discourage/offend project owners of those commands to have their own catalog
 
...those downsides and in particular the latter two is what has kept me away from doing it until now - but I'm also feeling we've tried for a few years now to get project owners to adopt it so something has to give.

Thus I'm inclined to say - lets go make jbangapp a thing. Yes, it won't be most efficient, it wont give stable versioning but it will make every awesome tool available more or less instantly.

For the latter part of possibly offending/discouraging @parttimenerd gave an idea...and i think *maybe* @quintesse proposed similar in the past:

# global aliases for catalogs

(sorry - need a better name)

This idea is that on top of jbang's ability to resolve catalogs available on github/gitlab/bitbucket (`jbang playwright@microsoft`) and DNS names (i.e. `jbang tree@xam.dk` refers to https://xam.dk/jbang-catalog.json)  we would have a new mechanism that jbang to have aliases looked up in another catalog.

ie. `jbang sqlline@@sqlline@` or `jbang jfr2flame@@ap-loader@` would be "expanded" according to a file bundled in jbang or even from github.com/jbangdev/jbang-catalog somewhere that defines whatever `@ap-loader@` shuold get expanded to.

It could be `jvm-profiling-tools/ap-loader` or it could be `jbangapp/jbang-catalog` ...and it could change over time.

This keeps the association to the project name (i.e. ap-loader) and makes it possible to make even shorter.

This global alias would be truly global - if users want custom aliases just use the existing alias mechanism; but with this we have the ability to offer an approach to spread the awesomeness of jbang while still recognizing project owners names/repo.

It could also backfire and many will be upset it does not make it clear jbangapp is the "owner" of it.

@parttimenerd suggested this could be alleviated by having jbang print out when it resolves @ap-line@ so its visible it comes from jbangapp or jvm-profiling-tools/ap-profiler.

WDYT ? Do one or the other or both?

p.s. I grabbed https://github.com/jbangapp already :)

**Comment by quintesse** [comment]
So the first part of this is clear to me and seems like a good idea.

It's the second proposal that I'm somewhat leery about. Messing about with (even more) magic symbols just seems very unintuitive. Personally I'm already not extremely happy with what we currently have (and I designed it) and this just seems to complicate it even more.

I understand it would be very easy to just be able to do `jbang jfr2flame@ap-loader` (or `jbang jfr2flame@@ap-loader@` and the like in case of expansion) but isn't that all a bit too much black magic? Of lookups upon lookups? We might get lost in a forest of lookups (we lookup a global alias to look up a catalog to look up an alias).

We can already name catalogs, you can for example say `jbang catalog add --name ap-loader http://url/to/catalog`. Wouldn't it be better to have people explicitly add a catalog they are interested in? A one-time command that's slightly longer but afterwards the aliases to run can use the shorter format.

We can already add catalogs to catalogs (I'm actually not really sure how we're using that or _if_ we're even using that as of today, so perhaps that's an opportunity) so imagine something like this:

```
$ jbang catalog list jbangapp
Catalogs:
----------
ap-loader = Commands for dealing with ap-loader
                    (https://github.com/jvm-profiling-tools/ap-loader/blob/main/jbang-catalog.json)
sqlline =     Some useful sqlline scripts
                   (https://raw.githubusercontent.com/jbangdev/jbangapp/master/sqlline/jbang-catalog.json)
.
.
.
$ jbang catalog add sqlline@jbangapp
$ jbang run@sqlline
```

(Btw, right now you can't add catalogs this way, that would need to be implemented)

This would allow for short aliases in daily use and still leave open the possibility that at some point the sqlline devs decide to support JBang directly, and a simple `jbang catalog add sqlline` would switch to the official catalog.

**Reply by quintesse** [reply]
And in that latter case where sqlline starts supporting JBang they could open a PR on the jbangapp catalog to update the sqlline entry and have it point to their official catalog. That way `jbang catalog add sqlline` and `jbang catalog add sqlline@jbangapp ` would have the same effect.

**Reply by parttimenerd** [reply]
This sounds like a great idea. Albeit it's more work for the users (I would have to tell them in my README to run the catalog add to use jbang and ap-loader easily).

**Reply by maxandersen** [reply]
I'm for sure not married to the @@ alias lookup notion - what I'm after is not have jbangapp catalog grow large/have too often updates; so for sure lets explore actually making catalog mentioning catalogs somewhat useful.

But asking to add a catalog is a downer; I'm after not having to tell users to add N catalogs to use a few commands. Today I can do `jbang sqlline@maxandersen` and `jbang byteman@byteman-project` or `jbang jfr2flame@jvm-profiling-tools/ap-profiler` with no extra steps. Its that level of easy we wanna have.

requiring `jbang catalog add sqlline@jbangapp` to then be able to run `jbang run@sqlline` is slower but also "polutes" the "well known" namespaces.

...but ..what about having aliases in catalogs that have been added to a catalog be visible?

i.e.  lets say we added this to jbangapp catalog:

```
"catalogs": {
    "maxandersen": {
      "catalog-ref": "https://github.com/maxandersen/jbang-catalog/blob/HEAD/jbang-catalog.json",
      "includeAliases": "true"
    }
  },
```

...now any alias in the maxandersen catalog would be available? 

also - we should be able to make it so just because catalog is listed in jbangapp it should not "redefine" maxandersen.
Maybe add a "expose": "false" flag? 

or simply have a toggle - that catalog referred in a catalog is either a "make alias for catalog" or "expose aliases" ...never both.

that wont solve the often update issue - but maybe we could optimize it and have aliases that refer to such aliases be installed/referenced directly to the catalog rather than via the parent catalog?

**Reply by quintesse** [reply]
>  Its that level of easy we wanna have.

Well... I guess it would be somewhat of a logical extension to say that if: `xxx@katalog` means the item "xxx" in catalog "katalog"... it follows that it could refer to a catalog named "xxx" as much as it could be an alias named "xxx". So perhaps we could extend our run syntax to allow:

```
jbang run@sqlline@jbangapp
````

Meaning the alias "xxx" in the catalog "yyy" defined in the catalog named "katalog" (and no, no more sub sub levels please :-) ).

> "includeAliases": "true"

Not sure I'm fond of this option, I don't think it should be the owner of the catalog that gets to decide if their commands get added to my local catalog but me, the user, with a kind of "--import" option or something like that.
`

**Reply by maxandersen** [reply]
What do you prefer? That everyone have to redeclare aliases or that catalogs can import other and re-export the aliases in a way you (and jbang) can trace where an alias actually refers to ?

Without this I have to keep duplicating catalogs or live with each user figuring list of catalogs they want. 

I'm not suggesting we remove the ability to control it as you already can but having to explicitly import catalogs is irrelevant here because any catalog can simply expose any alias and you don't control that.

I think it's much worse than adding a catalog today globally redefine catalog names. 

Ie. 'jbang catalog add evilrepo' can redefine what 'jbang myapp@maxandersen' refers to. 

 Would much rather have the ability to be able to express that means evilrepo contain the same aliases as maxandersen. 

Ie. What I'm saying is that Todays defaults is surprising and leads to loss of control by the user and too much power to the creator of catalogs.

**Reply by quintesse** [reply]
@maxandersen was this a reply to the last thing I wrote? Because I explicitly replied to the fact that you wanted something "easier" and suggested being able to run `jbang run@sqlline@jbangapp` without having to add any repos :-)

**Reply by maxandersen** [reply]
I get that - I'm asking why the catalog name used in importing the catalog *must* be used ? 

If that's the case i still need to manually duplicate the aliases. 

I think it's interesting to allow it as an option but I prefer being able to not have imported catalogs affect users own catalog names - which is currently the case.

**Reply by aalmiray** [reply]
s/jbangapp/jbangcentral/ 😏 

What if the central catalog were to be split as an index for other catalogs? Such that central contains name/url pairs such as

```json
"index": [
  {
    "sqlline":  "https://github.com/maxandersen/sqlline-jbang-catalog/blob/HEAD/jbang-catalog.json"
  },
  ...
]
```

The referenced `jbang-catalog.json` has the following content:

```json
{
  "aliases" : {
    "sqlline" : {
      "script-ref" : "sqlline.java",
      "description" : "Do the thing"
    }
  }
}
```

When invoking `jbang sqlline@@sqlline` the following should happen:

1. Yes, that's a double `@` that implies lookup on `jbangcentral`. value before `@@` is the alias on the target catalog, value after `@@` is the index name in the central catalog. This allows executing more than one alias per index name. Perhaps `jbang sqlline@@` could be used as a short hand for both alias and index name.
2. jbang checks if its central catalog is up to date
3. next, finds a match for the target index name. Fail if no match. If match checks if referenced catalog is up to date
4. now that it has a copy of the referenced catalog perform the same checks as before that is, match the given alias

**Reply by maxandersen** [reply]
not a super fan of "central" ...is that because of maven ccentral you suggest that?

jbangapp is nice and short :)

also, jbangapp is a github org so just realizing could literally just do sqlline@jbangapp/sqlline  ... could of course still have jbang sqlline@jbangapp as a shorter "catch all" place.

**Reply by aalmiray** [reply]
> not a super fan of "central" ...is that because of maven ccentral you suggest that?

Bingo!

**Reply by maxandersen** [reply]
@aalmiray your idea if i grok it would be possible even without a "special" index - if we make jbang sqlline@@sqlline`  be short hand for `jbang sqlline@jbangapp/sqlline` and `jbangapp` is the "default" alias for @. 

Maybe even make it sqlline@/sqlline ...i.e. empty implicit git org lookup...

@quintesse wdyt about that ? I don't think that "conflicts" what we have today.

**Reply by maxandersen** [reply]
Now if we did that (i.e. `sqlline@/sqlline`) I would still prefer there was a way to have jbangapp/sqlline/jbang-catalog go point to sqlline own catalog if that exists some day. having it just import the aliases, no "catalog aliasing" as what is happening today.

wdyt @quintesse ?

**Reply by maxandersen** [reply]
turned out it was easy to do...we already fail on "empty catalog" so turned that into mean "jbangapp"  https://github.com/jbangdev/jbang/pull/1738

wdyt?

**Reply by maxandersen** [reply]
> > not a super fan of "central" ...is that because of maven ccentral you suggest that?
> 
> Bingo!
> 

Going with jbanghub. Not specific to app; and hub is shorter than central :)

**Comment by maxandersen** [comment]
As an experiment I started to add content to https://github.com/jbanghub.

Thoughts made from chatting with @quintesse was that we would default import this catalog so all its catalogs would be available to users. Kinda felt like how we would have done appstore from the start if we knew then what we know today...

Idea was to make it so sqlline@sqlline would end up resoling to sqlline@jbanghub/sqlline and if sqlline one day made their own we could just point to it.

But after adding content to jbanghub I've kinda pulled back from it - because its hard to name things :) 

i.e. sqlline is actually not in github.com/sqlline but in github.com/julianhyde/sqlline - so sqlline catalog name is kinda globally namesquatting sqlline....but also it removes julianhyde...and maybe julianhyde likes that is how to refer to...so we could add julianhyde/sqlline alias - but then it starts looking like its him who manages it ... of course julianhyde here is just used as an example, it could be same with orgs/corporations like apple, microsoft, eclipse etc.

Thus for now I've gone with NOT having jbanghub use its power of default imported and instead added things that are in jbanghub/* org as `jbanghub/...` names.

This means there is actual value for people to add their own jbang-catalog even if jbanghub has something that maps to it, if they do that there shorthand name wins automatically.

For example, I added jbanghub/eclipse - it has ecj and transformer cli in it at this stage. 
It will be listed as `jbanghub/eclipse` in catalog (because default imported) and command will be `ecj@jbanghub/eclipse`.
If eclipse adds their own - `jbang ecj@eclipse` will work without interference.

...and if we do this - we could still add the notion of no catalog means jbanghub...so `jbang ecj@/eclipse` will be available for shorthand form.

This gives consistency and control to the original authors but jbang community can also easily add commands and have them visible.

wdyt - am I overthinking it ? and should we do the shorthand form or just stay with enabling jbanghub as a "thing"?

**Reply by maxandersen** [reply]
related bug I found with alias naming of catalogs not handled consistently: https://github.com/jbangdev/jbang/issues/1746
